### PR TITLE
Add support for casting `polars` array-like input to `numpy`

### LINKED ISSUE #5516: Add support for pandas and polars dataframes
### Describe the feature you would like to be added.

Polars and especially Pandas are very core Python data processing libraries by now. It would be nice to be able to create a mesh and/or add scalars directly from these datasets instead of the (very costly) casting to Numpy arrays.
Now the Polydata creation raises: 
```
TypeError: Invalid Input type:

Expected first argument to be either a:
- vtk.PolyData
- pyvista.PolyData
- numeric numpy.ndarray (1 or 2 dimensions)
- List (flat or nested with 3 points per vertex)
- vtk.vtkDataArray

Instead got: <class 'polars.series.series.Series'>
```

And the scalar assigning takes ridiculously long time (did not manage to see if it works in the end)

### Links to VTK Documentation, Examples, or Class Definitions.

_No response_

### Pseudocode or Screenshots

```python
import pyvista as pv
import polars as pl
import numpy as np

N = int(1e6)
data = pl.DataFrame(dict(
    points = np.random.sample((N, 3)),
    color = np.random.sample((N, 4))
    ))

cloud = pv.PolyData(data["points"])
cloud["colors"] = data["color"]
```

### PR DESCRIPTION
**Pull request by user27182**
This PR modifies `cast_to_ndarray` to attempt to gracefully deal with array-like inputs from `polars`. It is a possible initial step to fix #5516.

### CODE REVIEW

**Comment by tkoyama010** [comment]
I haven't followed discussing #5516 well, but can we support pandas too?

**Comment by user27182** [comment]
The TLDR of #5516 is that `pandas` and `polars` have array-like data structures which are implicitly supported anywhere `np.asarray` is used. Not only that, but calling `np.asarray` likely does not result in any copies of the data being made through the entire `pandas` (or `polars`) -> `numpy` -> `vtk` chain since they all make use of numpy's array APIs. The catch, though, is that the data must be contiguous.  And `polars` does not store data in its data frames contiguously by default, so calling `np.asarray` on a `polars` data frame will raise an error.

This PR is effectively a workaround for the `polars` error that is generated.

Another possible issue with supporting `pandas` was mentioned in #265, where using `.values` was needed to make use of `pandas` data. However, this is no longer the case, and using `.values` is [not recommended](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.values.html). Calling `np.asarray` will implicitly call `to_numpy` on the data and just work.

The other issue mentioned in #265 is that calling `np.asarray` will happily create object arrays. However, this case is already dealt with by `cast_to_ndarray`. The function will still allow string arrays, though, so I suppose we could add a dtype check for `np.number` types only (if we want to support complex numbers) or `np.floating` and `np.integer`. If we want to make this vtk-safe we should not allow complex numbers since [`numpy_to_vtk`](https://github.com/Kitware/VTK/blob/1df5e2c0629462569ac491a54729514f26db8256/Wrapping/Python/vtkmodules/util/numpy_support.py#L104) does not handle complex arrays.

**If we add the dtype check then I think this PR would effectively support both `polars` and `pandas`.**

The next steps (future PRs) to support these types would be to either:
1. replace `isinstace(x, np.array)` checks in pyvista code with `isinstace(x, _SupportsArray)` (see [#5516 (comment)](https://github.com/pyvista/pyvista/issues/5516#issuecomment-1915608738)) and then call `cast_to_ndarray(x)`, or
2. remove all instance checks and simply call `cast_to_ndarray(x)` on array-like inputs.

`cast_to_ndarray(x)` is also used by pyvista's validation package, so calling `validate_array(x)` would also work in case other validation checks are needed.

**Comment by user27182** [comment]
Another quirk with `polars` is that after casting, the ndarray has its writable flag set to False by default. This is fine for plotting but could pose an issue for any pyvista methods that mutate arrays directly. 

This could just be a safe default choice by `polars` and may be ok to change this to True. `polars` is written in Rust so this kind of safe practice may stem more from a design philosophy rather than a technical requirement. Even with the flag, it may be possible for vtk to mutate the array anyway after calling `numpy_to_vtk` since vtk will access the data at the c level. 

The impact of keeping the `writable=False` has not been tested.

**Comment by darikg** [comment]
Just a big picture comment here, but I wonder if it might be worth waiting for #5266 and do more integration of the validation package and an overall review of locations that check input arrays?

- `PolyData.__init__`
Checks `isinstance(var_inp, (np.ndarray, list, _vtk.vtkDataArray))` and calls `vtk_points` which does a bunch of custom array validation that should be entirely replaced by by one the validation functions

- `CellArray.__init__`
Simply calls `np.asarray` and then `pv.core.utilities.cells.numpy_to_idarr`, which could be replaced by one of the validation methods

- `DataSetAttributes.set_array` 
Calls DataSetAttributes.`_prepare_array` which does a lot of custom validation

- `core.utilities.arrays._coerce_pointslike_arg` 
Used in a bunch of locations, could be removed and replaced by `validate_arrayNx3` I think

etc. I'd feel a lot better about adding more complexity in the form of e.g. Polars support if it was all isolated to the validation package and the rest of pyvista used the validation more uniformly. E.g. when are we supposed to use `cast_to_ndarray` from the `core.utilities.arrays` package instead of `validate_array` in the validation package?

**Comment by darikg** [comment]
> And polars does not store data in its data frames contiguously by default, so calling np.asarray on a series which is part of polars data frame will raise an error.

I thought it just silently copied the data?

**Comment by user27182** [comment]
> it might be worth waiting for #5266 and do more integration of the validation package and an overall review of locations that check input arrays?

Yes I agree. Clean up the input validation across pyvista first, then add support for more general array-like inputs later.

> when are we supposed to use `cast_to_ndarray` from the `core.utilities.arrays` package instead of `validate_array` in the validation package?

It does seem a bit unclear. `cast_to_ndarray` was added specifically for input validation as part of #5100. It was added to the `pyvista.core.utilities.arrays` namespace at the time since it seemed to have general use. However, it may be best to move it directly into the `pyvista.core.validation` namespace, e.g. as part of #5226, and make it private and mostly just for internal use as part of `validate_array`.

**Comment by darikg** [comment]
> However, it may be best to move it directly into the pyvista.core.validation namespace, e.g. as part of https://github.com/pyvista/pyvista/pull/5226, and make it private and mostly just for internal use as part of validate_array.

That sounds good! Right now it's not possible for pyvista.core.utilities.arrays to import from validation because of the circular import

**Comment by user27182** [comment]
> > And polars does not store data in its data frames contiguously by default, so calling np.asarray on a series which is part of polars data frame will raise an error.
> 
> I thought it just silently copied the data?

In my view, `polars` _should_ silently copy the data. Unfortunately it does not. This code will fail:

``` python
import polars as pl
import numpy as np
data = pl.DataFrame(dict(points=[[1, 2, 3], [4, 5, 6]]))
array_in = np.asarray(data['points'])
```
``` polars.exceptions.ComputeError: 'to_numpy' not supported for dtype: List(Int64) ```

The workaround in this PR does exactly that though and will create a copy instead, although not silently (a warning is issued).

This case is covered by this PR here.
https://github.com/pyvista/pyvista/blob/8fc55b5e80b983eb1af97b11d765d4a9a66218ed/tests/core/test_utilities.py#L983-L991

IMHO I think this is surprising behaviour and may be worth raising an issue on the `polars` github. `polars` should probably be catching the exception and printing a warning directly.

**Comment by darikg** [comment]
Oh! Thanks for explaining. I agree that does seem like a miss on polars' part.

**Review by tkoyama010 — DISMISSED** [review]

**Review by tkoyama010 — APPROVED** [review]

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
Thanks. Let's wait!
